mm2src/coins/tx_history_db.rs

```diff
@@ -0,0 +1,443 @@
+                InitDbError::EmptyTableList
+                | InitDbError::DbIsOpenAlready { .. }
+                | InitDbError::InvalidVersion(_)
+                | InitDbError::OpeningError(_)
+                | InitDbError::TypeMismatch { .. }
+                | InitDbError::UnexpectedState(_)
+                | InitDbError::UpgradingError { .. } => TxHistoryError::InternalError(e.to_string()),
+            }
+        }
+    }
+
+    impl From<DbTransactionError> for TxHistoryError {
+        fn from(e: DbTransactionError) -> Self {
+            match e {
+                DbTransactionError::ErrorSerializingItem(_) => TxHistoryError::ErrorSerializing(e.to_string()),
+                DbTransactionError::ErrorDeserializingItem(_) => TxHistoryError::ErrorDeserializing(e.to_string()),
+                DbTransactionError::ErrorUploadingItem(_) => TxHistoryError::ErrorSaving(e.to_string()),
+                DbTransactionError::ErrorGettingItems(_) => TxHistoryError::ErrorLoading(e.to_string()),
+                DbTransactionError::ErrorDeletingItems(_) => TxHistoryError::ErrorClearing(e.to_string()),
+                DbTransactionError::NoSuchTable { .. }
+                | DbTransactionError::ErrorCreatingTransaction(_)
+                | DbTransactionError::ErrorOpeningTable { .. }
+                | DbTransactionError::UnexpectedState(_)
+                | DbTransactionError::TransactionAborted
+                | DbTransactionError::NoSuchIndex { .. }
+                | DbTransactionError::InvalidIndex { .. } => TxHistoryError::InternalError(e.to_string()),
+            }
+        }
+    }
+
+    type LoadHistoryResult = TxHistoryResult<Vec<TransactionDetails>>;
+    type SaveHistoryResult = TxHistoryResult<()>;
+    type ClearHistoryResult = TxHistoryResult<()>;
+
+    #[derive(Debug)]
+    enum TxHistoryEvent {
+        LoadHistory {
+            ticker: String,
+            result_tx: oneshot::Sender<LoadHistoryResult>,
+        },
+        SaveHistory {
+            ticker: String,
+            txs: Vec<TransactionDetails>,
+            result_tx: oneshot::Sender<SaveHistoryResult>,
+        },
+        Clear {
+            ticker: String,
+            result_tx: oneshot::Sender<ClearHistoryResult>,
+        },
+    }
+
+    pub struct TxHistoryDb {
+        event_tx: mpsc::Sender<TxHistoryEvent>,
+    }
+
+    #[async_trait]
+    impl TxHistoryOps for TxHistoryDb {
+        async fn init_with_fs_path(_path: PathBuf) -> TxHistoryResult<TxHistoryDb> {
+            let (init_tx, init_rx) = oneshot::channel();
+            let (event_tx, event_rx) = mpsc::channel(1024);
+
+            Self::init_and_spawn(init_tx, event_rx);
+            init_rx.await.expect_w("The init channel must not be closed")?;
+            Ok(TxHistoryDb { event_tx })
+        }
+
+        async fn load_history(&mut self, ticker: &str, _wallet_address: &str) -> LoadHistoryResult {
```

Comment:
Seems like if the users will (re) log in using another seed they will still see the transaction history of the previous address as `_wallet_address` is not used. Please consider using an ID built from ticker *and* address instead of the `ticker` field in TxHistoryEvent variants.

Reply:
Thanks for the important comment! I've changed the code, please check it again :)